### PR TITLE
Epmlstrcmw 49 Common interface for AuthMode

### PR DESCRIPTION
**Pull request by TimurKustov**
Created a common interface for AuthMode(still refactoring)

### CODE REVIEW

File: cloudSupport/src/main/scala/cromwell/cloudsupport/AuthMode.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package cromwell.cloudsupport
+
+import com.google.api.client.json.jackson2.JacksonFactory
+import com.google.auth.Credentials
+import cromwell.cloudsupport.AuthMode.OptionLookup
+import org.slf4j.LoggerFactory
+
+object AuthMode {
+
+  class OptionLookupException(val key: String, cause: Throwable) extends RuntimeException(key, cause)
```
Comment (iliaonishchenko) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/AuthMode.scala]: It is best practice to extend `Exception with Product with Serializable` for custom exceptions

File: cloudSupport/src/main/scala/cromwell/cloudsupport/AuthMode.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  type OptionLookup = String => String
+  type CredentialsValidation = Credentials => Unit
+
+  lazy val jsonFactory = JacksonFactory.getDefaultInstance
+
+}
+
+trait AuthMode[A] {
+
+  lazy val log = LoggerFactory.getLogger(getClass.getSimpleName)
```
Comment (iliaonishchenko) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/AuthMode.scala]: Why do we define logger in trait?

File: cloudSupport/src/main/scala/cromwell/cloudsupport/AwsAuthMode.scala [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package cromwell.cloudsupport
+
+import software.amazon.awssdk.auth.credentials.AwsCredentials
+
+abstract class AwsAuthMode extends AuthMode[AwsCredentials] {
+
+  val _credential: AwsCredentials
```
Comment (iliaonishchenko) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/AwsAuthMode.scala]: Is it notation which is used in Cromwell?

File: cloudSupport/src/main/scala/cromwell/cloudsupport/AwsModes.scala [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+import software.amazon.awssdk.services.sts.StsClient
+import software.amazon.awssdk.services.sts.model.AssumeRoleRequest
+
+
+object AwsModes {
+
+  import CredentialValidatorInstances.awsCredentialValidation
+  import CredentialValidatorSyntax._
+
+  case object MockAuthMode extends AwsAuthMode {
```
Comment (iliaonishchenko) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/AwsModes.scala]: Why do we have mock in business logic code?
Is it used somewhere in the project except tests?

File: cloudSupport/src/main/scala/cromwell/cloudsupport/AwsModes.scala [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+  object CustomKeyMode
+
+  final case class CustomKeyMode(override val name: String,
+                                 accessKey: String,
+                                 secretKey: String,
+                                 region: Option[String]
+                                ) extends AwsAuthMode {
+
+    override lazy val _credential: AwsCredentials = {
```
Comment (iliaonishchenko) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/AwsModes.scala]: Why is it lazy?

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AuthMode.scala [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  type CredentialsValidation = Credentials => Unit
+
+  lazy val jsonFactory = JacksonFactory.getDefaultInstance
+
+}
+
+trait AuthMode[A] {
+
+  lazy val jsonFactory = JacksonFactory.getDefaultInstance
+  lazy val log = LoggerFactory.getLogger(getClass.getSimpleName)
```
Comment (mnesvit) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AuthMode.scala]: Could you, please, explain, why should logger be lazy?

**Review by iliaonishchenko — CHANGES_REQUESTED** [review]
Nice job, Timur!

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+import software.amazon.awssdk.services.sts.model.AssumeRoleRequest
+
+object AwsModes {
+
+  import CredentialValidatorInstances.awsCredentialValidation
+  import CredentialValidatorSyntax._
+
+  case object MockAuthMode extends AwsAuthMode {
+
+    override val _credential: AwsCredentials = AnonymousCredentialsProvider.create.resolveCredentials()
```
Comment (svitkovsergey) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala]: Please consider using camelCase when naming variables

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+  final case class CustomKeyMode(override val name: String,
+                                 accessKey: String,
+                                 secretKey: String,
+                                 region: Option[String]
+                                ) extends AwsAuthMode {
+
+    override lazy val _credential: AwsCredentials = {
+
+      val storage = Creds.apply(Option(AwsBasicCredentials.create(accessKey, secretKey)), Option(null), region)
```
Comment (svitkovsergey) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala]: Please avoid `null` usage, use Option.None instead

**Review by svitkovsergey — CHANGES_REQUESTED** [review]
Big request! 👍
I left some comments after a brief review, I will look in more detail later

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala (hunk 8 above)
Comment (TimurKustov) [reply cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala]: Thank you, a good point.

File: cloudSupport/src/main/scala/cromwell/cloudsupport/AwsModes.scala (hunk 4 above)
Comment (TimurKustov) [reply cloudSupport/src/main/scala/cromwell/cloudsupport/AwsModes.scala]: No, it isn't.

File: cloudSupport/src/main/scala/cromwell/cloudsupport/AuthMode.scala (hunk 2 above)
Comment (TimurKustov) [reply cloudSupport/src/main/scala/cromwell/cloudsupport/AuthMode.scala]: Good point! Defining logger at abstract class for google modes would be a better decision.

File: cloudSupport/src/main/scala/cromwell/cloudsupport/AuthMode.scala (hunk 1 above)
Comment (TimurKustov) [reply cloudSupport/src/main/scala/cromwell/cloudsupport/AuthMode.scala]: Do you mean it is better to use case classes?

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AuthMode.scala (hunk 6 above)
Comment (TimurKustov) [reply cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AuthMode.scala]: From my answer to @ilyaonishenko, it is more appropriately to define logger at GoogleAuthMode class. So then lazy val seems like a good because it used only once at one of the authentification google modes.

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsAuthMode.scala [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package cromwell.cloudsupport.auth
+
+import cromwell.cloudsupport.auth.AuthMode.OptionLookup
+import software.amazon.awssdk.auth.credentials.AwsCredentials
+
+abstract class AwsAuthMode extends AuthMode[AwsCredentials] {
+
+  val _credential: AwsCredentials
+
+  override def credentials(options: OptionLookup): AwsCredentials
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsAuthMode.scala]: What is the point of this override?

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+    override val _credential: AwsCredentials = AnonymousCredentialsProvider.create.resolveCredentials()
+
+    override val name = "no_auth"
+
+    override def credentials(options: OptionLookup): AwsCredentials = _credential
+
+  }
+
+  object CustomKeyMode
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala]: Is there any reason for this object to exist?

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    override lazy val _credential: AwsCredentials = {
+      // The ProfileCredentialsProvider will return your [default]
+      // credential profile by reading from the credentials file located at
+      // (~/.aws/credentials).
+      //
+      // Validate credentials synchronously here, without retry.
+      // It's very unlikely to fail as it should not happen more than a few times
+      // (one for the engine and for each backend using it) per Cromwell instance.
+      Creds.apply(Option(DefaultCredentialsProvider.create.resolveCredentials()), Option(null), region).validate
+    }
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala]: You can place the comment at the top of the `_credential` lazy val and remove unnecessary curly braces. 

Also, I'm really confused with laziness. Why do you make `_credential` lazy in some places, while in some other places it's a regular value?

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+      val requestBuilder = AssumeRoleRequest
+        .builder
+        .roleSessionName("cromwell")
+        .roleArn(roleArn)
+        .durationSeconds(3600)
+
+      // The builder is simply mutating itself (TODO: find good ref, as v2
+      // uses generated code)
+      // So we can get away with a val and discard the return value
+      if (!externalId.isEmpty) requestBuilder.externalId(externalId)
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala]: use `externalId.nonEmpty`

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        case _ => throw new RuntimeException(s"Base auth object has already been assigned")
+      }
+    }
+
+    // We want to allow our tests access to the value
+    // of the baseAuthObj
+    def baseAuthentication(): AuthMode[AwsCredentials] = {
+      baseAuthObj match {
+        case Some(o) => o
+        case _ => throw new RuntimeException(s"Base auth object has not been set")
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala]: Can be rewritten as `baseAuthObj.getOrElse(throw new ...)`

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+        stsCredentials.accessKeyId,
+        stsCredentials.secretAccessKey,
+        stsCredentials.sessionToken)
+
+      Creds.apply(Option(sessionCredentials), Option(null), region).validate
+    }
+
+    override def credentials(options: OptionLookup): AwsCredentials = _credential
+
+    private var baseAuthObj: Option[AuthMode[AwsCredentials]] = None
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala]: A `var`, seriously?

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/CredentialValidator.scala [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+          builder.credentialsProvider(StaticCredentialsProvider.create(credentials.get))
+            .build
+            .getCallerIdentity(GetCallerIdentityRequest.builder.build)
+          ()
+        }
+
+      override def validateCredentials(credentials: Creds): AwsCredentials = {
+        val awsCredentials = Creds.unapply(credentials).get._1
+        val region = Creds.unapply(credentials).get._3
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/CredentialValidator.scala]: Uh
Oh
What the hell is going on here?
1) Do not use `.get` on an Option, especially without check whether it is empty
2) N_E_V_E_R use methods like `._1`, `._2` etc on tuples.

In your case, you can achieve the desired result by typing
```val Creds(awsCredentials, _, region) = credentials```

Same applies to other places in this PR

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/GoogleAuthMode.scala [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+    credentials(GoogleAuthMode.NoOptionLookup, Nil)
+  }
+
+  /**
+   * Alias for credentials(options, Nil).
+   * Only valid for credentials that are NOT externally provided and do not need scopes, such as ApplicationDefault.
+   */
+  override def credentials(options: OptionLookup): OAuth2Credentials = {
+    credentials(options, Nil)
+  }
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/GoogleAuthMode.scala]: Unnecessary curly braces

**Review by myazinn — CHANGES_REQUESTED** [review]

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala (hunk 10 above)
Comment (TimurKustov) [reply cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala]: nobody dies if i remove it

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsAuthMode.scala (hunk 9 above)
Comment (TimurKustov) [reply cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsAuthMode.scala]: same as CustomKeyMode object

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala (hunk 11 above)
Comment (TimurKustov) [reply cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala]: At some cases (for ex. at AssumeRoleMode) laziness used as decision to avoid forward references. But anyway it is unneccessary and much better to move specified fields to the top of the class.

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala (hunk 14 above)
Comment (TimurKustov) [reply cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala]: It hurts eyes, but still this var reassigns at 103, any suggestions what to do with that?

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/CredentialValidator.scala (hunk 15 above)
Comment (TimurKustov) [reply cloudSupport/src/main/scala/cromwell/cloudsupport/auth/CredentialValidator.scala]: Thank you! It is realy serious defect. I will take it into consideration.

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AuthMode.scala [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AuthMode.scala]: Remove this extra space

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AuthMode.scala [hunk 18]
```diff
@@ -0,0 +1,35 @@
+package cromwell.cloudsupport.auth
+
+import com.google.api.client.json.jackson2.JacksonFactory
+import com.google.auth.Credentials
+import cromwell.cloudsupport.auth.AuthMode.OptionLookup
+G
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AuthMode.scala]: What is this? Does this code even compile?

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+
+
+  final case class AssumeRoleMode(override val name: String,
+                                  baseAuthName: String,
+                                  roleArn: String,
+                                  externalId: String,
+                                  region: Option[String]
+  ) extends AwsAuthMode {
+
+    private var baseAuthObj: Option[AuthMode[AwsCredentials]] = None
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala]: I made some research and it looks like it is possible to get rid of this `var`.
However, it is debatable whether it should be done in this PR.

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+import cromwell.cloudsupport.auth.AuthMode.OptionLookup
+import software.amazon.awssdk.auth.credentials._
+import software.amazon.awssdk.regions.Region
+import software.amazon.awssdk.services.sts.StsClient
+import software.amazon.awssdk.services.sts.model.AssumeRoleRequest
+
+object AwsModes {
+
+  import CredentialValidatorInstances.awsCredentialValidation
+  import CredentialValidatorSyntax._
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala]: Why not import it outside the object?

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsAuthMode.scala [hunk 21]
```diff
@@ -0,0 +1,9 @@
+package cromwell.cloudsupport.auth
+
+import software.amazon.awssdk.auth.credentials.AwsCredentials
+
+abstract class AwsAuthMode extends AuthMode[AwsCredentials] {
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsAuthMode.scala]: Previously it was a sealed trait. Why did you change trait to an abstract class and put all of its implementations into a separate object?

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsAuthMode.scala [hunk 22]
```diff
@@ -0,0 +1,9 @@
...
+
+  val credential: AwsCredentials
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsAuthMode.scala]: `_credential` in the original code is private most of the time. Maybe you should make this `val` in you code at least `protected`?

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/CredentialValidator.scala [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+object CredentialValidatorSyntax {
+
+  implicit class CredentialValidatorSyntax[A](credentials: Creds) {
+    def validate(implicit v: CredentialValidator[A]): A = {
+      v.validateCredentials(credentials)
+    }
+  }
+}
+
+
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/CredentialValidator.scala]: remove this extra line

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/GoogleAuthMode.scala [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport
+import com.google.api.client.http.HttpResponseException
+import com.google.auth.http.HttpTransportFactory
+import com.google.auth.oauth2.OAuth2Credentials
+import cromwell.cloudsupport.auth.AuthMode.OptionLookup
+import org.slf4j.LoggerFactory
+
+import scala.util.{Failure, Success, Try}
+
+abstract class GoogleAuthMode extends AuthMode[OAuth2Credentials] {
```
Comment (myazinn) [inline cloudSupport/src/main/scala/cromwell/cloudsupport/auth/GoogleAuthMode.scala]: Same question about why this is an abstract class now

**Review by myazinn — CHANGES_REQUESTED** [review]
Much better! However, there are some things to improve

File: cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala (hunk 19 above)
Comment (TimurKustov) [reply cloudSupport/src/main/scala/cromwell/cloudsupport/auth/AwsModes.scala]: Okay, i will correct this, if it will be required.

**Comment by Mikhail-Sosnin** [comment]
What is the current state of this request?
Do we need to close it?

**Comment by Mikhail-Sosnin** [comment]
@TimurKustov 
What is the current state of this request?
Do we need to close it?

**Comment by Mikhail-Sosnin** [comment]
@TimurKustov ?

**Comment by mnesvit** [comment]
outdated
